File: src/main/java/org/poglebiarka/app/notifications/google/NotificationClient.java
package org.poglebiarka.app.notifications.google;

import org.poglebiarka.app.notifications.Link;
import org.poglebiarka.app.notifications.NotificationRequest;
import org.springframework.web.client.RestClient;

import static org.springframework.http.MediaType.APPLICATION_JSON;

public class NotificationClient {

    final RestClient restClient;

    public NotificationClient(RestClient restClient) {
        this.restClient = restClient;
    }

    public void sendForAll(NotificationRequest request) {
        restClient.post()
                .uri("/fcm/send")
                .contentType(APPLICATION_JSON)
                .body(toGoogleNotificationRequest(request))
                .retrieve()
                .toBodilessEntity();
    }

    private GoogleNotificationRequest toGoogleNotificationRequest(NotificationRequest request) {
        return new GoogleNotificationRequest(
                "/topics/all",
                "high",
                new NotificationHead(
                        request.mainTitle(),
                        request.mainDescription(),
                        "default"),
                new NotificationData(
                        request.subTitle(),
                        request.subDescription(),
                        request.links().stream().map(this::toNotificationButton).toList()
                )
        );
    }

    private NotificationButton toNotificationButton(Link link) {
        return new NotificationButton(link.title(), link.url());
    }
}

File: src/main/kotlin/org/poglebiarka/admin/notifications/google/NotificationClient.kt
package org.poglebiarka.admin.notifications.google

import org.poglebiarka.admin.notifications.NotificationRequest
import org.springframework.http.MediaType
import org.springframework.web.client.RestClient

class NotificationClient(val restClient: RestClient) {

    fun sendForAll(request: NotificationRequest) {
        restClient.post()
            .uri("/fcm/send")
            .contentType(MediaType.APPLICATION_JSON)
            .body(GoogleNotificationRequest.from(request))
            .retrieve()
            .toBodilessEntity()
    }

}
